SwiftCompilerSources/Sources/Optimizer/Utilities/OwnershipLiveness.swift

```diff
@@ -0,0 +1,977 @@
+  //   b1 = borrow d1
+  //   br bb3(b1)
+  // bb2:
+  //   b2 = borrow d1
+  //   br bb3(b2)
+  // bb3(reborrow):
+  //   u1 = d1
+  //   u2 = reborrow
+  //   // can't move destroy above u2
+  //   destroy_value d1
+  //
+  // Dominating definingValue example: walkDown must continue visiting
+  // uses of a guaranteed phi in the outer lifetime:
+  //
+  // bb0:
+  //  b1 = borrow d1
+  //  cond_br bb1, bb2
+  // bb1:
+  //   p1 = projection b1
+  //   br bb3(p1)
+  // bb2:
+  //   p1 = projection b1
+  //   br bb3(p2)
+  // bb3(forwardingPhi):
+  //   u1 = b1
+  //   u2 = forwardingPhi
+  //   // can't move end_borrow above u2
+  //   end_borrow b1
+  private mutating func walkDown(guaranteedPhi: Phi) -> WalkResult {
+    guard visited.insert(guaranteedPhi.value) else {
+      return .continueWalk
+    }
+    var enclosingValues = Stack<Value>(context)
+    defer { enclosingValues.deinitialize() }
+    gatherEnclosingValues(for: guaranteedPhi.value, in: &enclosingValues,
+                          context)
+    guard enclosingValues.contains(definingValue) else {
+      // Since definingValue is not an enclosing value, it must be
+      // consumed or reborrowed by some outer adjacent phi in this
+      // block. An outer adjacent phi's uses do not contribute to the
+      // outer liveness. Instead, guaranteedPhi will be recorded as a
+      // regular lifetime-ending use by the visitor.
+      return .continueWalk
+    }
+    // definingValue is not consumed or reborrowed by an outer
+    // adjacent phi in guaranteedPhi's block. Therefore this
+    // guaranteedPhi's uses contribute to the liveness of
+    // definingValue.
+    //
+    // TODO: instead of relying on Dominance, we can reformulate
+    // this algorithm to detect redundant phis, similar to the
+    // SSAUpdater.
+    if !definingValue.parentBlock.dominates(guaranteedPhi.successor,
+      context.dominatorTree) {
+      // definingValue does not dominate guaranteedPhi. Record this
+      // unenclosed phi so the liveness client can insert the missing
+      // outer adjacent phi.
+      unenclosedPhis.append(guaranteedPhi);
+      return .continueWalk
+    }
+    // Since definingValue dominates guaranteedPhi, this is a well-formed linear
+    // lifetime, and liveness can proceed.
+    if guaranteedPhi.isReborrow {
+      return visitUsesOfInner(value: guaranteedPhi.value)
+    } else {
+      return visitUsesOfOuter(value: guaranteedPhi.value)
+    }
+  }
+}
+
+/// Visit all uses of an interior address that keep the parent object alive.
+///
+/// See C++ TransitiveAddressWalker.
+///
+/// Requires:
+///   mutating func leafUse(address:) -> WalkResult
+///   mutating func dependent(value:dependsOn:) -> WalkResult
+///   mutating func pointerEscape(address:) -> WalkResult
+///   mutating func unknown(address:) -> WalkResult
+///
+/// TODO: Verify that pointerEscape is only called for live ranges in which
+/// `findPointerEscape()` returns true.
+protocol AddressUseVisitor {
```

Comment:
That Walker convention is good but I don't want to rename AddressUseVisitor yet. Its purpose of is to visit all of the address uses, not just the leaf uses, partitioning them by the kind of use. This is the most general possible address use visitor, so there's no other sensible name.

Note that I originally wrote `AddressUseVisitor` as an extension of `AddressDefUseWalker`, even though it's obviously backward, just to keep up the appearance of reuse. It was complex and bug-prone, and the results were invalid.

For now, I need to separate this new infrastructure from the code that affects the rest of the compiler. So redesigning `AddressDefUseWalker` should be a separate step. In fact, the separate list of opcodes is necessary until we want to change the behavior of existing passes.

Literally the only thing you want to reuse here is the list of address projection opcodes. The simple, obvious way to do that is with an AddressProjection abstraction. That would solve any short term problem. In general, forcing inheritance to reuse logic leads to incomprehensible, buggy code.

Longer term, the two utilities should be integrated. There are multiple problems with the current design and it's not possible to write a simple address use visitor on top of it. One thing I'll point out is that `AddressDefUseWalker.leafUse` is just outright wrong. It's called on just any unrecognizable use, and that's not suitable for utilities that are designed to be complete.

`AddressDefUseWalker` should be built on top of `AddressUseVisitor`. It should simply override `projection` to update the path and it should be responsible for the def-use walk itself, rather than just visiting the use. Then AddressUseVisitor won't have to do any walk and everything will make sense and work naturally.

Reply:
> The simple, obvious way to do that is with an AddressProjection abstraction.

AddressDefUseWalker _is_ the address projection abstraction. Its implementation mainly consists of a switch over all address projection instructions and "translating" that to a projection path.

Reply:
I'll use `AddressLifetimeDefUseWalker` for this. Later, I think the other one should be renamed`AddressPathDefUseWalker` or `AddressProjectionDefUseWalker`. We should still add an `AddressProjection` protocol.